gelf_writer.go

```diff
@@ -30,6 +30,9 @@ type Writer struct {
 	Facility         string // defaults to current process name
 	CompressionLevel int    // one of the consts from compress/flate
 	CompressionType  CompressType
+
+	gzipWriter *gzip.Writer
```

Comment:
Since these fields are not exported, why don't we keep a single `zw` (`io.WriteCloser`) field instead?

Reply:
`Writer.CompressionType` is exported, so to have a single "current writer" field would require checking that the writer type in that field is OK for the current `CompressionType` value. Something along the lines of changing `if w.gzipWriter == nill` to `if w.zw == nill and w.zwCompressionType == w.CompressionType`.

If that is the approach you prefer I'll add it.

Reply:
Taking another look at the code, `Writer.CompressionLevel` is also exported, so that needs to be checked before re-using a writer as well. Since storing a writer for every possible compression level isn't a good idea, I think `if w.zw == nill and w.zwCompressionType == w.CompressionType and w.zwCompressionLevel == w.CompressionLevel` is what we need.

I tried refactoring `bufferedWriter` into a proper `WriterCloser` (including `Reset`) when I created this PR but couldn't get that to work. Since checking for `CompressionType` and `CompressionLevel` add more repetitive work (as does the `err` check you mentioned), I'll give that another go.